src/backend/defs/go-objects/device.go

```diff
@@ -0,0 +1,275 @@
+		d.DateRevoked, _ = time.Parse(time.RFC3339Nano, revokedAt.(string))
+	}
+	if status, ok := input["status"].(string); ok {
+		d.Status = status
+	}
+	if i_type, ok := input["type"].(string); ok {
+		d.Type = i_type
+	}
+	if userAgent, ok := input["user_agent"].(string); ok {
+		d.UserAgent = userAgent
+	}
+	if userId, ok := input["user_id"].(string); ok {
+		if id, err := uuid.FromString(userId); err == nil {
+			d.UserId.UnmarshalBinary(id.Bytes())
+		}
+	}
+
+	return nil
+}
+
+func (d *Device) UnmarshalJSON(b []byte) error {
+	input := map[string]interface{}{}
+	if err := json.Unmarshal(b, &input); err != nil {
+		return err
+	}
+
+	return d.UnmarshalMap(input)
+}
+
+// return a JSON representation of Device suitable for frontend client
+func (d *Device) MarshalFrontEnd() ([]byte, error) {
+	return JSONMarshaller("frontend", d)
+}
+
+// bespoke implementation of the json.Marshaller interface
+// outputs a JSON representation of an object
+// this marshaller takes account of custom tags for given 'context'
+func (d *Device) JSONMarshaller() ([]byte, error) {
+	return JSONMarshaller("", d)
+}
+
+func (d *Device) NewEmpty() interface{} {
+	nd := new(Device)
+	nd.Locations = DeviceLocations{}
+	return nd
+}
+
+func (d *Device) MarshallNew(args ...interface{}) {
+	if len(d.DeviceId) == 0 || (bytes.Equal(d.DeviceId.Bytes(), EmptyUUID.Bytes())) {
+		d.DeviceId.UnmarshalBinary(uuid.NewV4().Bytes())
+	}
+	if len(d.UserId) == 0 || (bytes.Equal(d.UserId.Bytes(), EmptyUUID.Bytes())) {
+		if len(args) == 1 {
+			switch args[0].(type) {
+			case UUID:
+				d.UserId = args[0].(UUID)
+			}
+		}
+	}
+	if d.DateInsert.IsZero() {
+		d.DateInsert = time.Now()
+	}
+	d.DateRevoked = time.Time{}
+
+	for i, _ := range d.Locations {
+		d.Locations[i].MarshallNew()
+	}
+
+	if strings.TrimSpace(d.Type) == "" {
+		d.Type = DefaultDeviceType()
+	}
+}
+
+func (d *Device) JsonTags() map[string]string {
+	return jsonTags(d)
+}
+
+func (d *Device) SortSlices() {
+	sort.Sort(ByIpAddress(d.Locations))
+}
+
+// GetRelatedList returns a map[PropertyKey]Type of structs that are embedded into a Device from joined tables
+func (d *Device) GetRelatedList() map[string]interface{} {
+	return map[string]interface{}{
+		"Locations": &DeviceLocation{},
+		//"PublicKey": &PublicKey{},
+	}
+}
+
+func (d *Device) GetSetRelated() <-chan interface{} {
+	getSet := make(chan interface{})
+	if d.Locker == nil {
+		d.Locker = new(sync.Mutex)
+	}
+	go func(*sync.Mutex, chan interface{}) {
+		d.Locker.Lock()
+
+		// send locations
+		for i, _ := range d.Locations {
+			getSet <- &(d.Locations[i])
+		}
+
+		// send publicKeys
+		for i, _ := range d.PublicKeys {
+			getSet <- &(d.PublicKeys[i])
+		}
+
+		// all done
+		close(getSet)
+		d.Locker.Unlock()
+	}(d.Locker, getSet)
+
+	return getSet
+}
+
+// IsValidDeviceType checks the `t` string against the constant `DeviceTypes`
+// returns true if `t` is a valid type for a device.
+func IsValidDeviceType(t string) bool {
```

Comment:
Is there not a way to define like an enumeration, instead of a delimited string ?

Reply:
no `enum` type in GO. See https://github.com/golang/go/issues/19814 and the response from lance taylor : « A new language feature needs compelling use cases. All language features are useful, or nobody would propose them; the question is: are they useful enough to justify complicating the language and requiring everyone to learn the new concepts? What are the compelling use cases here? How will people use these? For example, would people expect to be able to iterate over the set of valid enum values, and if so how would they do that? Does this proposal do more than let you avoid adding default cases to some switches? »

Reply:
So forget enum name, think a map or a structure of valid values. Not a single string like this.

Reply:
Could be a `map[string]bool`, but map are unordered ; Thus there is no way with a map to find the `default` string for device.  
Also, making it a simple string makes it easy to put it in a config file, easying edit of devices' types list.

Reply:
Changed to an array in commit below